Add support to limit the what gets downstreamed in v0

tmshort opened the pull request:
The merge tool always pulls the master branch from all the staging repos.
This is a problem if one of the dependencies has been updated to a version
of something that is incompatible with the base repository, and the base
repository still references an older dependency.

For instance: operator-lifecycle-manager may reference older verisons
of operator-framework/api and operator-framework/operator-registry.

If one of those dependencies is updated to a more recent version of say,
k8s, but OLM hasn't pulled it in, that will cause merge problems.

What should happen is upstream OLM should integrate with those newer
dependencies before they're pulled downstream. Until then, the dependencies
should remain at the versions listed in OLM's depdencies, as specified in
go.mod.

This is done by looking at the latest OLM and seeing what versions
of `api` and `operator-registry` should be referenced.

In addition to the above:

Remove the `go mod` commands run in staging, those commands are not
necessary as upstream should be running them and keeping their go.*
files maintained.

Add logic to handle commits that have been downstreamed after the
tagged version that OLM expects. This is mostly for the transision
period to limiting downstreaming by tag.

stevekuznetsov commented on pkg/v0/cmd.go:
```diff
@@ -160,9 +181,86 @@ func Run(ctx context.Context, logger *logrus.Logger, opts Options) error {
+	return "", fmt.Errorf("%q not found", repo)
+}
+
+func calculateRepoRefs(ctx context.Context, logger *logrus.Entry, opts Options) (map[string]string, error) {
```
Why don't we re-use the logic in the v1 bumper for this?

tmshort replied:
Was there logic? Let me look.

stevekuznetsov replied:
The bits that query `go.mod` should be re-usable at least!

tmshort replied:
Yup, I just did it a bit different. Using `go list` ought to be more reliable.

stevekuznetsov commented on pkg/v0/cmd.go:
```diff
@@ -224,19 +305,33 @@ func detectNewCommits(ctx context.Context, logger *logrus.Entry, stagingDir, cen
+		if len(strs) != 4 {
+			return nil, fmt.Errorf("unexpected command output %q", output)
+		}
+		fetchRef := strs[2]
```
Why? The _whole point_ of `FETCH_HEAD` is to refer to what you just fetched. This adds more logic to the script and the best you can do is correctly parse the `--porcelain` output to get what `FETCH_HEAD` already refers to. Please don't do this.

tmshort replied:
I could be mis-remembering, but I recall there was a situation where `git fetch` didn't actually fetch anything, so `FETCH_HEAD` wasn't updated. AFAIR, this was not as much an issue running the bumper as a job, but running the bumper locally.

tmshort replied:
After I cleared out the old remotes that the old scripts had used, it seems to be better behaved without this.

stevekuznetsov commented on pkg/v0/cmd.go:
```diff
@@ -224,19 +305,33 @@ func detectNewCommits(ctx context.Context, logger *logrus.Entry, stagingDir, cen
-		if _, err := internal.RunCommand(repoLogger, exec.CommandContext(ctx,
+		ref, ok := repoRefs["operator-framework/" + repo]
+		if !ok {
+			logger.Debugf("using 'master' for repo %q", repo)
```
Prefer to be explicit. Do explicit defaulting in the caller. If `repo` is not in `repoRefs` here, returning an error would be appropriate.

tmshort commented:
@stevekuznetsov this should fix all your concerns

everettraven commented on pkg/v0/cmd.go:
```diff
@@ -160,9 +169,94 @@ func Run(ctx context.Context, logger *logrus.Logger, opts Options) error {
 
+func getTagOrCommit(ctx context.Context, repo string, opts Options, logger *logrus.Entry) (string, error) {
+	module := fmt.Sprintf("github.com/%s", repo)
+	rawInfo, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+		"go", "list", "-json", "-m", module), "staging/operator-lifecycle-manager"))
```
No very familiar with this repo so I'm trying to make sure I understand the flow that is taken here. 

With this using the `staging/operator-lifecycle-manager` directory assumes that the upstream OLM `master` change has already been downstreamed or it would result in the dependency repos being behind by one sync.

The way I'm currently understanding this is that the upstream OLM `master` branch will be synced followed by the api + o-reg branches/versions on the next sync - is this accurate?

tmshort replied:
You actually make a good point, and we haven't been able to test this scenario. The operator-lifecycle-manager repo is fetched, and then the other repos are fetched - based on the current state of this directory. Nothing has been merged/synced. So out-of-date data _could_ be used. Then the OLM changes are synced without the corresponding API or operator-registry changes, and things blow up.

So, I _think_ what needs to happen is:
1. Fetch the OLM repo.
2. Look for new commits on the OLM repo.
3. Merge the next commit from the OLM repo.
4. fetch the API and operator-registry repos based on `staging/operator-lifecycle-manager`
5. Merge those API and operator-registry commits.
6. Run `go mod` commands
7. Run `make manifest`
8. Goto step 3.

Because, now that I think about it, the updates from API and operator-registry are ignored because `go list` is used at the wrong time.

joelanford commented on pkg/v0/cmd.go:
```diff
@@ -26,6 +27,11 @@ const (
 	githubRepo = "operator-framework-olm"
 )
 
+var depRepos = []string{
+	"operator-framework/api",
+	"operator-framework/operator-registry",
+}
+
```
Nit: define this somewhere non-global. Maybe in the Options struct? Or maybe in the `Run` command, and then pass through function signatures.

tmshort replied:
This is used in exactly one place in the code, and could simply be moved there? Although I like this _type_ of stuff at the beginning so it can be more easily modified (which we should never have to do _here_).

tmshort commented:
So, based on @everettraven [comment](https://github.com/openshift/operator-framework-tooling/pull/19#discussion_r1540039584), and my response, I need to rethink the integration.

tmshort commented:
I think I figured out a way to get the versions of `api` and `operator-registry` that need to sync with `operator-lifecycle-manager`, without having to sync/merge OLM commits first... this will be easier to implement, and keep things inline with the current logic and operation.

tmshort commented:
I added logic to better handle the new api & operator-registry versions that OLM wants, and this occurs before any commits are cherry-picked.
I also added logic to handle the transition to label-based vs latest downstreaming by not erroring out when we've already downstreamed "too much" (which is the current status)
Finally, removed the `go mod` commands within the `staging` directory, as it's upstream's responsibility to make sure they are keeping `go.mod` tidy (and with @grokspawn latest operator-registry commit, they are tidy)

everettraven commented on pkg/v0/cmd.go:
```diff
@@ -160,9 +171,127 @@ func Run(ctx context.Context, logger *logrus.Logger, opts Options) error {
+		return nil, err
+	}
+	defer os.RemoveAll(dir)
+	if _, err := internal.RunCommand(logger, exec.CommandContext(ctx,
+		"git", "worktree",
+		"add",
+		dir,
+		"FETCH_HEAD",
+	)); err != nil {
+		return nil, err
+	}
```
TIL about `git worktree` - neat!

Do we need to be concerned about any branches this might create on a local machine? Reading the [documentation on `git worktree`](https://git-scm.com/docs/git-worktree) it looks like we could use the `-d` flag to create a "throwaway" worktree (although maybe this doesn't matter at all since we attempt to clean up the directory later).

tmshort replied:
Yes, and the neat thing is that it can be _any_ reference in the repo, including things you fetched from outside of the origin.
When run as a job, it's done in a temporary container, so cleanup isn't critical.
_But_ if you run it locally, it will still show up in `git worktree list`, and using the `git worktree prune` will cleanup any number of them. In my case, I had many left over from testing.

stevekuznetsov replied:
I would recommend not doing this. We have many years of experience with automated tooling that uses `git` and not storing references in filesystem state (branches, etc) is a less error-prone approach. That way, your program has no expectations about the environment it's run in and it can be correct regardless of what kind of `git` state exists in a container, on a dev machine, whatever. Plus, no cleanup necessary.

everettraven commented:
Just to make sure I have the correct understanding now of the flow, we essentially:
- Fetch o-f/olm `master`
- "Checkout" the `master` branch (`FETCH_HEAD`) into a temporary directory
- Run dependency version analysis in the temp directory
- Clean up the temp directory
- Pull all the refs

If, at a high-level, I have a correct understanding of the flow this LGTM

tmshort commented:
> Just to make sure I have the correct understanding now of the flow, we essentially:
> 
> * Fetch o-f/olm `master`
> * "Checkout" the `master` branch (`FETCH_HEAD`) into a temporary directory
> * Run dependency version analysis in the temp directory
> * Clean up the temp directory
> * Pull all the refs
> 
> If, at a high-level, I have a correct understanding of the flow this LGTM

Yes.

tmshort commented:
Running this code in your local repo will cause the remote reference to be created that you may not want.
```
git reflog expire --expire-unreachable=now --all
git gc --prune=now
```
will clean up any of the fetched remotes.

everettraven commented:
> Running this code in your local repo will cause remote references to be created that you may not want.
> 
> ```
> git reflog expire --expire-unreachable=now --all
> git gc --prune=now
> go worktree prune
> ```
> 
> will clean up any cruft. This doesn't matter so much when run by the job in a container.

Can we document this somewhere more concrete with this change? Maybe the README?

everettraven commented:
>Maybe the README?

Just noticed this repo doesn't have a README 🤦 - might be worth adding

tmshort commented:
> > Maybe the README?
> 
> Just noticed this repo doesn't have a README 🤦 - might be worth adding. I'm okay with that being a follow up though

Yup, but that's probably better for another PR.

tmshort commented:
Readme: #25

tmshort commented:
ping?

everettraven commented:
/approve

grokspawn commented:
/approve

stevekuznetsov commented:
@tmshort for the comment in https://github.com/openshift/operator-framework-tooling/pull/19#discussion_r1469849388 ... why did we end up re-implementing logic that already existed in v1?

tmshort commented:
> @tmshort for the comment in [#19 (comment)](https://github.com/openshift/operator-framework-tooling/pull/19#discussion_r1469849388) ... why did we end up re-implementing logic that already existed in v1?

The logic is _very_ close but not exact. Also, I did not want to disrupt v1, as _at the time_ it was working. I didn't want to add risk.

stevekuznetsov commented:
@tmshort can you explain how the logic is different? It is quite literally identical behavior in parsing `go.mod` ...
